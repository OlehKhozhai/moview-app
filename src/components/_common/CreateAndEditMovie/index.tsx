import React from 'react';
import { Form, Formik, FormikHelpers, FormikProps } from 'formik';
import { useDispatch } from 'react-redux';

import Input from 'components/_common/FormFields/Input';
import Button from 'components/_common/Button';
import Select from 'components/_common/FormFields/Select';
import { createMovieAction, editMovieAction } from 'redux/actions';
import { Movie } from 'redux/types';
import { GENRES } from 'config';
import { createMovieValidationSchema } from 'helpers/validators';
import { getErrors } from 'helpers/common';
import styles from './styles.module.scss';

const defaultValues: Partial<Movie> = {
  title: '',
  release_date: '',
  poster_path: '',
  genres: [],
  overview: '',
  runtime: 0,
};

type CreateAndEditMovieProps = {
  action?: 'create' | 'edit';
  movieId?: number;
  initialValues?: Partial<Movie>;
  onSubmit: () => void;
};

const CreateAndEditMovie: React.FC<CreateAndEditMovieProps> = ({
  action = 'create',
  initialValues,
  movieId,
  onSubmit,
}) => {
  const dispatch = useDispatch();
  const formTitle = action === 'create' ? 'Add movie' : 'Edit movie';
  const initialFormValues = initialValues || defaultValues;

  const handleSubmit = async (
    values: Partial<Movie>,
    { setSubmitting, setErrors }: FormikHelpers<Partial<Movie>>
  ) => {
    let errors = null as unknown;

    setSubmitting(true);

    if (action === 'edit' && movieId) {
      errors = await dispatch<unknown>(editMovieAction({ ...values, id: movieId }));
    } else {
      errors = await dispatch<unknown>(createMovieAction(values));
    }

    if (!errors) {
      onSubmit();
      setSubmitting(false);
    } else {
      setErrors(getErrors(errors as string[]));
    }
  };

  return (
    <Formik
      initialValues={initialFormValues}
      validationSchema={createMovieValidationSchema}
      onSubmit={handleSubmit}
    >
      {({ isSubmitting }: FormikProps<Partial<Movie>>) => (
        <Form className={styles.root} data-testid="form">
          <h2 className={styles.title}>{formTitle}</h2>

          <Input name="title" label="Title" placeholder="Enter title" />
          <Input
            name="release_date"
            label="Release date"
            type="date"
            placeholder="Enter realise date"
          />
          <Input name="poster_path" label="Movie url" placeholder="Enter url" />
          <Select name="genres" label="Genre" placeholder="Choose Genres" options={GENRES} />
          <Input name="overview" label="Overview" placeholder="Enter overview" />
          <Input name="runtime" label="Runtime" placeholder="Enter runtime" type="number" />

          <div className={styles.buttonsWrapper}>
            <Button title="Reset" variation="primaryOutline" className={styles.resetBtn} />
            <Button
              title={action === 'create' ? 'Submit' : 'Save'}
              type="submit"
              disabled={isSubmitting}
            />
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default React.memo(CreateAndEditMovie);
