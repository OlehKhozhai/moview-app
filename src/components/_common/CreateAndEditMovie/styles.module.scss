@import 'assets/styles/mixins/colors';

.root {
  color: $primary-color;
}

.title {
  text-transform: uppercase;
  font-size: 40px;
  text-align: left;
  margin-bottom: 50px;
}

.buttonsWrapper {
  display: flex;
  justify-content: flex-end;
  margin-top: 30px;
}

.resetBtn {
  margin-right: 15px;
}
