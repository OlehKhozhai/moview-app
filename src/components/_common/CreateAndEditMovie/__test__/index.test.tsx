import React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react';

import CreateAndEditMovie from '../index';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn().mockImplementation(() => () => undefined),
}));

jest.mock('components/_common/FormFields/Input', () => ({
  __esModule: true,
  default: () => '---Input---',
}));
jest.mock('components/_common/FormFields/Select', () => ({
  __esModule: true,
  default: () => '---Select---',
}));

describe('CreateAndEditMovie component', () => {
  test('should render markup by default', () => {
    const component = create(<CreateAndEditMovie onSubmit={() => undefined} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render edit movie markup', () => {
    const component = create(
      <CreateAndEditMovie
        onSubmit={() => undefined}
        initialValues={{ title: 'test title' }}
        action="edit"
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should trigger onSubmit function', async () => {
    const onSubmitStub = jest.fn();
    const { queryByTestId } = render(
      <CreateAndEditMovie
        onSubmit={onSubmitStub}
        initialValues={{
          title: 'Test title',
          release_date: '2020',
          poster_path: 'poster_path',
          genres: ['Documentary'],
          overview: 'Overview',
          runtime: 10,
        }}
        action="edit"
      />
    );

    await waitFor(() => {
      const form = queryByTestId('form');

      if (form) {
        fireEvent.submit(form);
      }
    });

    expect(onSubmitStub).toHaveBeenCalled();
  });
});
