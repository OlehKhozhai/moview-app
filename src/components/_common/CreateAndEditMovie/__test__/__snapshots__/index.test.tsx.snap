// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAndEditMovie component should render edit movie markup 1`] = `
<form
  action="#"
  className="root"
  data-testid="form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <h2
    className="title"
  >
    Edit movie
  </h2>
  ---Input---
  ---Input---
  ---Input---
  ---Select---
  ---Input---
  ---Input---
  <div
    className="buttonsWrapper"
  >
    <button
      className="root primaryOutline resetBtn"
      disabled={false}
      type="button"
    >
      Reset
    </button>
    <button
      className="root primary"
      disabled={false}
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;

exports[`CreateAndEditMovie component should render markup by default 1`] = `
<form
  action="#"
  className="root"
  data-testid="form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <h2
    className="title"
  >
    Add movie
  </h2>
  ---Input---
  ---Input---
  ---Input---
  ---Select---
  ---Input---
  ---Input---
  <div
    className="buttonsWrapper"
  >
    <button
      className="root primaryOutline resetBtn"
      disabled={false}
      type="button"
    >
      Reset
    </button>
    <button
      className="root primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
